### PR TITLE
Create SharedmemRecording

### PR DESCRIPTION
**Pull request by samuelgarcia**
Implement a SharedMemoryRecording

This replace #2351 

this depend on #2363

### CODE REVIEW

File: src/spikeinterface/core/__init__.py [hunk 1]
```diff
@@ -79,14 +79,14 @@
...
-    write_to_h5_dataset_format,
-    write_binary_recording,
     read_python,
     write_python,
 )
 from .job_tools import ensure_n_jobs, ensure_chunk_size, ChunkRecordingExecutor, split_job_kwargs, fix_job_kwargs
 from .recording_tools import (
+    write_binary_recording,
+    write_to_h5_dataset_format,
+    write_binary_recording,
```
Comment (yger) [inline src/spikeinterface/core/__init__.py:89]: There is a double import to write_binary_recording here

**Comment by yger** [comment]
Don't you think that we could do a check for the total memory available while saving to memory, and prevent a call to save_to_memory() if the memory is not enough?

File: src/spikeinterface/core/tests/test_baserecording.py [hunk 2]
```diff
@@ -162,15 +162,21 @@ def test_BaseRecording():
     rec3 = BaseExtractor.load(cache_folder / "simple_recording")
 
     # cache to memory
-    rec4 = rec3.save(format="memory")
-
+    rec4 = rec3.save(format="memory", sahred=False)
```
Comment (yger) [inline src/spikeinterface/core/tests/test_baserecording.py:165]: typo here, should be shared

**Review by yger — COMMENTED** [review]
Looks good to me, thanks a lot for the effort :-)

**Review by yger — APPROVED** [review]
Looks good to me, minor some typos highlighted

File: src/spikeinterface/core/numpyextractors.py [hunk 3]
```diff
@@ -83,6 +84,22 @@ def __init__(self, traces_list, sampling_frequency, t_starts=None, channel_ids=N
             "sampling_frequency": sampling_frequency,
         }
 
+    @staticmethod
+    def from_recording(source_recording, **job_kwargs):
+        traces_list, shms = write_memory_recording(source_recording, dtype=None, **job_kwargs)
+        if shms[0] is not None:
+            # if the computation was done in parrralel then traces_list is shared array
```
Comment (alejoe91) [inline src/spikeinterface/core/numpyextractors.py:91]:
```suggestion
            # if the computation was done in parrallel then traces_list is shared array
```

File: src/spikeinterface/core/numpyextractors.py [hunk 4]
```diff
@@ -99,6 +116,115 @@ def get_traces(self, start_frame, end_frame, channel_indices):
...
+        List of sharedmem names.
+    shape_list: list
+
+    sampling_frequency: float
+        The sampling frequency in Hz
+    t_starts: None or list of float
+        Times in seconds of the first sample for each segment
+    channel_ids: list
+        An optional list of channel_ids. If None, linear channels are assumed
+    """
```
Comment (zm711) [inline src/spikeinterface/core/numpyextractors.py:139]:
```suggestion
    channel_ids: list, default: None
        An optional list of channel_ids. If None, linear channels are assumed
    t_starts: None or list of float, default: None
        Times in seconds of the first sample for each segment
    main_shm_owner: bool, default: True
    
    """
```

add in shape_list def and `main_shm_owner` def

File: src/spikeinterface/core/numpyextractors.py [hunk 5]
```diff
@@ -99,6 +116,115 @@ def get_traces(self, start_frame, end_frame, channel_indices):
...
+    extractor_name = "SharedMemory"
+    mode = "memory"
+    name = "SharedMemory"
+
+    def __init__(
+        self, shm_names, shape_list, dtype, sampling_frequency, channel_ids=None, t_starts=None, main_shm_owner=True
+    ):
+        
+        
+        assert len(shape_list) == len(shm_names)
```
Comment (zm711) [inline src/spikeinterface/core/numpyextractors.py:150]:
```suggestion
        assert len(shape_list) == len(shm_names), 'Each shm_name in `shm_names` must have a shape in `shape_list`'
```

File: src/spikeinterface/core/numpyextractors.py [hunk 6]
```diff
@@ -99,6 +116,115 @@ def get_traces(self, start_frame, end_frame, channel_indices):
...
+        assert all(shape_list[0][1] == shape[1] for shape in shape_list)
+
+        # create traces from shraedmem names
```
Comment (zm711) [inline src/spikeinterface/core/numpyextractors.py:153]:
```suggestion
        # create traces from sharedmem names
```

File: src/spikeinterface/core/numpyextractors.py [hunk 7]
```diff
@@ -99,6 +116,115 @@ def get_traces(self, start_frame, end_frame, channel_indices):
...
+
+        if channel_ids is None:
+            channel_ids = np.arange(traces_list[0].shape[1])
+        else:
+            channel_ids = np.asarray(channel_ids)
+            assert channel_ids.size == traces_list[0].shape[1]
+        BaseRecording.__init__(self, sampling_frequency, channel_ids, dtype)
+
+        if t_starts is not None:
+            assert len(t_starts) == len(traces_list), "t_starts must be a list of same size than traces_list"
```
Comment (zm711) [inline src/spikeinterface/core/numpyextractors.py:170]:
```suggestion
            assert len(t_starts) == len(traces_list), "t_starts must be a list of same size as traces_list"
```

**Review by zm711 — COMMENTED** [review]
docstring comments

File: src/spikeinterface/core/tests/test_recording_tools.py [hunk 8]
```diff
@@ -149,16 +149,16 @@ def test_write_memory_recording():
...
-    write_memory_recording(recording, dtype=None, verbose=True, n_jobs=1, chunk_memory="100k", progress_bar=True)
+    traces_list, shms = write_memory_recording(recording, dtype=None, verbose=True, n_jobs=1, chunk_memory="100k", progress_bar=True)
 
-    if platform.system() != "Windows":
-        # write parrallel
-        write_memory_recording(recording, dtype=None, verbose=False, n_jobs=2, chunk_memory="100k")
-
-        # write parrallel
-        write_memory_recording(recording, dtype=None, verbose=False, n_jobs=2, total_memory="200k", progress_bar=True)
+    # write parrallel
```
Comment (alejoe91) [inline src/spikeinterface/core/tests/test_recording_tools.py:158]:
```suggestion
    # write parallel
```

**Review by alejoe91 — APPROVED** [review]

File: src/spikeinterface/core/numpyextractors.py (hunk 4 above)
Comment (alejoe91) [reply src/spikeinterface/core/numpyextractors.py]: done :)

**Pull request merged**